File: Milestone1/python/core/coordinator.py
import asyncio
from dataclasses import dataclass
from typing import Dict, Any, List, Tuple
from core.buffer import RingBuffer
from core.acquisition import AcquisitionScheduler
from core.uploader import Uploader

@dataclass(frozen=True)
class Event:
    kind: str  # "POLL_TICK" | "UPLOAD_TICK"

class Coordinator:
    """
    Event-driven controller matching the updated Petri net:
    - Central Idle
    - Explicit Poll Ready / Upload Ready
    - Guards: Not Uploading, Not Polling (no overlap)
    """

    def __init__(self,
                 queue: asyncio.Queue,
                 buffer: RingBuffer,
                 acq: AcquisitionScheduler,
                 upl: Uploader):
        self.q = queue
        self.buffer = buffer
        self.acq = acq
        self.upl = upl
        # places (booleans)
        self.idle = True
        self.polling = False
        self.uploading = False
        self.poll_ready = False
        self.upload_ready = False

    async def run(self):
        while True:
            ev = await self.q.get()
            if ev.kind == "POLL_TICK":
                self.poll_ready = True
                print("[Poll Timer = 2s] tick → Poll Ready")
            elif ev.kind == "UPLOAD_TICK":
                self.upload_ready = True
                print("[Upload Timer = 15s] tick → Upload Ready")
            await self._drain_enabled_transitions()

    async def _drain_enabled_transitions(self):
        made_progress = True
        while made_progress:
            made_progress = False

            # Prefer upload to reduce latency when data exists
            if (self.upload_ready and not self.polling and not self.uploading
                    and self.buffer.not_empty()):
                print("Not Polling → Uploading")
                await self._do_uploading()
                made_progress = True
                continue

            if (self.poll_ready and not self.uploading and not self.polling):
                print("Not Uploading → Polling")
                await self._do_polling()
                made_progress = True
                continue

    async def _do_polling(self):
        self.idle = False
        self.polling = True

        ok, sample = await self.acq.poll_once()
        if not ok:
            print("Poll skipped (device unavailable)")
            self.polling = False
            self.idle = True
            return

        print("Sample Ready", sample)
        pushed = self.buffer.push(sample)
        if pushed:
            print("Buffer Push")
            print(f"Pushed | Buffer size = {self.buffer.size}")
        else:
            print("Buffer full — sample dropped")

        self.poll_ready = False
        self.polling = False
        self.idle = True
        print("Idle started")

    async def _do_uploading(self):
        self.idle = False
        self.uploading = True

        batch: List[Dict[str, Any]] = self.buffer.drain_all()
        print(f"Uploading (sending {len(batch)} samples)")
        ok = await self.upl.upload_once(batch)

        if ok:
            print("Received ACK → Idle")
            self.upload_ready = False
        else:
            print("Upload failed — re-queue for next window")
            for item in batch:
                self.buffer.push(item)
            self.upload_ready = False

        self.uploading = False
        self.idle = True

File: Milestone1/cpp/src/core/application/Coordinator.cpp
/**
 * @file Coordinator.cpp
 * @brief Implementation of the Coordinator class for managing acquisition and upload events.
 *
 * @details
 * Implements the Coordinator class methods, handling the main event-processing loop,
 * state transitions, polling, and uploading of inverter samples. Coordinates between
 * the event queue, ring buffer, acquisition scheduler, and uploader to manage system
 * state and ensure reliable operation.
 *
 * @author Yasith
 * @author Prabath
 * @version 1.0
 * @date 2025-08-18
 *
 * @par Revision history
 * - 1.0 (Yasith, 2025-08-18) Moved implementations to cpp file and split to layers.
 */


#include "Coordinator.h"

/**
 * @fn Coordinator::Coordinator
 * @brief Construct a Coordinator with required references to system components.
 *
 * @param q [in] Reference to the event queue.
 * @param buffer [in] Reference to the ring buffer storing samples.
 * @param acq [in] Reference to the acquisition scheduler for polling samples.
 * @param upl [in] Reference to the uploader for sending samples.
 * @param running [in] Reference to a shared atomic flag controlling execution.
 *
 * @note All referenced objects must outlive the Coordinator instance.
 */
Coordinator::Coordinator(BlockingQueue<Event>& q,
            RingBuffer<Sample>& buffer,
            AcquisitionScheduler& acq,
            Uploader& upl,
            std::atomic<bool>& running)
    : q_(q), buffer_(buffer), acq_(acq), upl_(upl), running_(running) {}

/**
 * @fn Coordinator::run
 * @brief Main event-processing loop for the Coordinator.
 *
 * @details
 * - Initializes internal state.
 * - Processes events from the queue until a shutdown condition is met.
 * - Responds to:
 *   - PollTick: Marks polling ready.
 *   - UploadTick: Marks uploading ready.
 *   - Shutdown: Terminates loop.
 */
void Coordinator::run() 
{
    idle_ = true; polling_ = false; uploading_ = false;
    poll_ready_ = false; upload_ready_ = false;

    for (;;) 
    {
        if (!running_.load()) break;
        Event ev = q_.pop();
        if (ev.kind == EventKind::Shutdown) break;

        if (ev.kind == EventKind::PollTick) 
        {
            poll_ready_ = true;
            std::cout << "[Poll Timer = 2s] tick -> Poll Ready\n";
        } 
        else if (ev.kind == EventKind::UploadTick) 
        {
            upload_ready_ = true;
            std::cout << "[Upload Timer = 15s] tick -> Upload Ready\n";
        }
        
        drain_enabled_transitions();
    }
}

/**
 * @fn Coordinator::request_stop
 * @brief Request the Coordinator to stop processing events.
 *
 * @details
 * Sets the shared `running_` flag to false, which will cause the main loop to exit.
 */
void Coordinator::request_stop()
{ 
    running_.store(false); 
}

/**
 * @fn Coordinator::drain_enabled_transitions
 * @brief Execute state transitions that are currently enabled.
 *
 * @details
 * - Checks for conditions to start uploading or polling.
 * - Continues looping while progress is made in state transitions.
 */
void Coordinator::drain_enabled_transitions() 
{
    bool progress = true;
    while (progress) 
    {
        progress = false;

        if (upload_ready_ && !polling_ && !uploading_ && buffer_.not_empty()) 
        {
            std::cout << "Not Polling -> Uploading\n";
            do_uploading();
            progress = true;
            continue;
        }
        
        if (poll_ready_ && !uploading_ && !polling_) 
        {
            std::cout << "Not Uploading -> Polling\n";
            do_polling();
            progress = true;
            continue;
        }
    }
}

/**
 * @fn Coordinator::do_polling
 * @brief Perform a polling operation to acquire a sample.
 *
 * @details
 * - Calls AcquisitionScheduler::poll_once to get a sample.
 * - If successful, stores the sample in the buffer.
 * - Handles buffer full condition.
 * - Updates internal state flags.
 */
void Coordinator::do_polling() 
{
    idle_ = false; polling_ = true;

    auto [ok, s] = acq_.poll_once();
    if (!ok) 
    {
        std::cout << "Poll skipped (device unavailable)\n";
        polling_ = false; idle_ = true; return;
    }

    std::cout << "Sample Ready {"
              << "'t': " << s.t << ", 'voltage': " << s.voltage
              << ", 'current': " << s.current << ", 'power': " << s.power
              << "}\n";

    bool pushed = buffer_.push(s);
    if (pushed) 
    {
        std::cout << "Buffer Push\n";
        std::cout << "Pushed | Buffer size = " << buffer_.size() << "\n";
    } 
    else 
    {
        std::cout << "Buffer full — sample dropped\n";
    }

    poll_ready_ = false; polling_ = false; idle_ = true;
    std::cout << "Idle started\n";
}

/**
 * @fn Coordinator::do_uploading
 * @brief Perform an upload operation for all buffered samples.
 *
 * @details
 * - Drains all samples from the buffer.
 * - Attempts to upload via Uploader::upload_once.
 * - On failure, requeues all samples back into the buffer.
 */
void Coordinator::do_uploading() 
{
    idle_ = false; uploading_ = true;

    auto batch = buffer_.drain_all();
    std::cout << "Uploading (sending " << batch.size() << " samples)\n";

    bool ok = upl_.upload_once(batch);
    if (ok) 
    {
        std::cout << "Received ACK -> Idle\n";
        upload_ready_ = false;
    } 
    else 
    {
        // CloudStub already printed failure; keep behavior consistent:
        for (const auto& s : batch) buffer_.push(s);
        upload_ready_ = false;
    }

    uploading_ = false; idle_ = true;
}
